### ISSUE TITLE
LoRA - Seed are not preserved on changing LoRA scale

### ISSUE THREAD

**Issue by elismasilva** [huggingface/diffusers#10165]
### Describe the bug

Hello everyone, I have a question, I noticed that when I use LoRA in diffusers the seeds are not preserved in all generations while I change the LoRA scale. When I do the same thing in auto 1111 this does not happen, to be honest it happens but very little variation

## Diffusers
|**Without Lora**| scale=0.5 | scale=1.0|
| :-----------: | :-----------: | :-----------: |
|<img src="https://github.com/user-attachments/assets/f229d2e7-8f1c-49d2-bd4d-d8249ae85a91" width=512/>|<img src="https://github.com/user-attachments/assets/9e58b26d-e472-4100-8645-51c53f6ee048" width=512/>|<img src="https://github.com/user-attachments/assets/47203e3f-5a7e-48d3-bb07-fc44944896d9" width=512/>|

## AUTO 1111
|**Without Lora**| scale=0.5 | scale=1.0|
| :-----------: | :-----------: | :-----------: |
|<img src="https://github.com/user-attachments/assets/c1933312-700f-4193-8b60-096b3de9c17c" width=512/>|<img src="https://github.com/user-attachments/assets/be73286f-f818-4a01-b649-aca0ba29c665" width=512/>|<img src="https://github.com/user-attachments/assets/e862862a-a59d-4369-a168-82cfd31ea507" width=512/>|

### Reproduction

```python
from diffusers import StableDiffusionXLPipeline
import diffusers
import torch

model_id="stabilityai/stable-diffusion-xl-base-1.0"
pipe = StableDiffusionXLPipeline.from_pretrained(model_id, 
                                         variant="fp16", 
                                         use_safetensors=True,
                                         torch_dtype=torch.float16).to("cuda")
pipe.enable_model_cpu_offload()

add_kwargs = {}
add_kwargs["use_karras_sigmas"] = True
add_kwargs["algorithm_type"] = "sde-dpmsolver++"
scheduler = getattr(diffusers, "DPMSolverMultistepScheduler")
pipe.scheduler = scheduler.from_config(pipe.scheduler.config, timestep_spacing='trailing',**add_kwargs)  

seed=867459680

generator = torch.Generator("cpu").manual_seed(seed)   
prompt = "absolutely monochrome white dog faded in white, white fog, white faded, pale white, clear white design, pale skin, white on white, white tones, maximum white, soft contrast"
negative_prompt = ""

with torch.no_grad():
    prompt_ = pipe.encode_prompt(prompt=prompt, 
                                negative_prompt=negative_prompt, 
                                do_classifier_free_guidance=True, 
                                num_images_per_prompt=1,                             
                                device="cuda")
prompt_params = {
    "prompt_embeds": prompt_[0],                
    "negative_prompt_embeds": prompt_[1],
    "pooled_prompt_embeds": prompt_[2],                
    "negative_pooled_prompt_embeds": prompt_[3],
}

scale=0.5
lora_path="models/LoRA/sd_xl_offset_example-lora_1.0.safetensors"
pipe.load_lora_weights(lora_path, torch_dtype=torch.float32, adapter_name="offset")
pipe.set_adapters(["offset"], adapter_weights=scale)
pipe.enable_lora()

img = pipe(   
    num_inference_steps=30,     
    guidance_scale=7, 
    generator=generator,    
    **prompt_params,
    ).images[0]

img.save("dog.png")

```

### Logs

```shell
N/A
```

### System Info

Copy-and-paste the text below in your GitHub issue and FILL OUT the two last points.

- 🤗 Diffusers version: 0.31.0.dev0
- Platform: Windows-10-10.0.19045-SP0
- Running on Google Colab?: No
- Python version: 3.10.11
- PyTorch version (GPU?): 2.4.0+cu121 (True)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Huggingface_hub version: 0.25.2
- Transformers version: 4.40.1
- Accelerate version: 0.29.3
- PEFT version: 0.12.0
- Bitsandbytes version: 0.43.1
- Safetensors version: 0.4.4
- xFormers version: 0.0.27.post2
- Accelerator: NVIDIA GeForce RTX 3060 Ti, 8192 MiB
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

 @sayakpaul

**Comment by sayakpaul** [comment]
I am not sure if your example conveys your issue correctly. 

scale of 0.5 and scale of 1.0 are bound to produce different results than no LoRA. 

If you are asking if they should produce the same result when the scale and seed are the same, then yes, they should. Maybe sure, `generator` is initialized each time before it's passed to the pipeline call. This is because its internal states are consumed in place. 

So, doing 

```py
generator = torch.manual_seed(0)

image_1 = pipeline(..., generator=generator)
image_2 = pipeline(..., generator=generator)
```

is not the same as doing:

```py
image_1 = pipeline(..., generator=torch.manual_seed(0))
image_2 = pipeline(..., generator=torch.manual_seed(0))
```

**Comment by FurkanGozukara** [comment]
@sayakpaul excellent answer

what  internal states it has that are being consumed could you give some examples?

**Comment by sayakpaul** [comment]
Yes, generator should be initialized always.

**Comment by elismasilva** [comment]
Thank you! so i will close this issue, apparently this was just a misunderstanding on how to use the generator correctly.

**Comment by asomoza** [comment]
just adding for future reference for the search, this is in the docs [here](https://huggingface.co/docs/diffusers/using-diffusers/reusing_seeds?hardware=CPU#control-randomness):

> It might seem unintuitive to pass Generator objects to a pipeline instead of the integer value representing the seed. However, this is the recommended design when working with probabilistic models in PyTorch because a Generator is a random state that can be passed to multiple pipelines in a sequence. As soon as the Generator is consumed, the state is changed in place which means even if you passed the same Generator to a different pipeline, it won’t produce the same result because the state is already changed.

**Comment by elismasilva** [comment]
> because a Generator is a random state that can be passed to multiple pipelines in a sequence

Is there any scenario where using the same generator would be useful? When I was testing the refiner, I passed the same generator to it and my image was changed, so I had to leave the refiner without the generator, but then I made the adjustment by restarting the generator and it worked. So, this passage in the text seems to imply that there are possible scenarios, the refiner's one is not the case since it also works without the generator. Unless the text is referring to the initial variable, for example:

```python
g=torch.Generator(device="cuda")
```
So in this case it would be the same generator variable, but then between pipelines these generators need to be reinitialized every time:

```python
#pipe 1
g.manual_seed(0) #Initialize
pipe(..., generator=g)

#pipe 2
g.manual_seed(0) #Initialize again
pipe(..., generator=g)
```

**Comment by asomoza** [comment]
I'm in the same boat as you, I thought the generator didn't change the state and that could be reused multiple times but it's not like that. I really don't know the use case for it to be used multiple times, probably this is rooted in some ML use case which we don't use here (probably for batches too).

In the documentation there's a load state you can use for using the same generator state each time but it's the same as just initializing it again with the same seed, don't know if there a computational difference between the two.

**Comment by elismasilva** [comment]
Yes, I even thought that maybe this is useful within the same pipeline when consumed internally it reuses the state in something that was running in a loop or something like that, however, at least for SDXL I didn't find anything other than the generation of the latent. Maybe some other pipeline makes good use of this feature.
